SLM UART refactoring (#11416)
nrfconnect/sdk-nrf: applications/serial_lte_modem/doc/slm_data_mode.rst

  * The TCP server is stopped.
  * The remote server disconnects the TCP client.
  * The TCP client disconnects from the remote server due to an error.
  * The UDP client disconnects from the remote server due to an error.

  Currently, these scenarios move the SLM application to state where the data received from UART is dropped, until the MCU sends the termination command.

Triggering the transmission
===========================

The SLM application buffers all the arbitrary data received from the UART bus before initiating the transmission.

The transmission of the buffered data to the LTE network is triggered:
  * By the time limit when the defined inactivity timer times out.
  * By the reception of the the termination string.
  * By the filling of the data mode buffer.

If there is no time limit configured, the minimum required value applies.
For more information, see the `Data mode control #XDATACTRL`_  command.

Flow control in data mode
=========================

When SLM fills its UART receive buffers, it enables the UART hardware flow control, which disables UART reception.
SLM reenables UART reception when the data has been moved to the data mode buffer.

> If the data mode buffer fills, the data will be transmitted to the LTE network.

Reviewer: ```suggestion
If the data mode buffer fills, the data are transmitted to the LTE network.
```

.. note::
   There is no unsolicited notification defined for this event.
   UART hardware flow control is responsible for imposing and revoking flow control.

The data mode buffer size is controlled by :ref:`CONFIG_SLM_DATAMODE_BUF_SIZE <CONFIG_SLM_DATAMODE_BUF_SIZE>`.

.. note::

>    The whole buffer will be sent in a single operation.
>    When transmitting UDP packets, only one complete packet should reside in the data mode buffer at any time.

Reviewer: ```suggestion
   The whole buffer is sent in a single operation.
   When transmitting UDP packets, only one complete packet must reside in the data mode buffer at any time.
```
